### PR TITLE
Remaining tweaks to middleware refactor + initial documentation changes for ``v7``

### PR DESCRIPTION
**Pull request by fselmo**
### What was wrong?

- Fix some inconsistencies in the middleware onion logic dealing with unhashable `Web3Builder.build` references.
- Remove references to middleware classes and opt for using the classes themselves.
- Allow turning off of ``HTTPProvider`` retries on exceptions if `exception_retry_configuration` is set to `None`; add test. Fix the default exception distinctions for sync vs async providers.
- Update the documentation for the middleware refactor, request retry configuration and caching.
- Started a ``v6`` -> ``v7`` migration guide.
- The ``AsyncHTTPProvider`` wasn't loading the right default exceptions for exception retry while I was documenting it. We also couldn't effectively turn off the retries for the sync provides. I went ahead and updated that in this PR as well and added the relevant test to make sure that is covered.
- ``normalize_request_parameters`` middleware was sitting in a stale state with no internal use. I removed it and updated the migration guide to reflect that.

Related to Issue #3169 

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

### CODE REVIEW

File: docs/examples.rst [hunk 1]
```diff
@@ -688,8 +688,8 @@ Inject the middleware into the middleware onion
 
 .. code-block:: python
 
-    from web3.middleware import extradata_to_poa_middleware
-    w3.middleware_onion.inject(extradata_to_poa_middleware, layer=0)
+    from web3.middleware import ExtradataToPOAMiddleware
```
Comment (pacrob) [inline docs/examples.rst]: consistency nit: the error types all start `ExtraData...` and the field in question is `extraData`. What'chu think about changing the middleware name to `ExtraData...`?

File: docs/v7_migration.rst [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+Breaking Changes:
+
+
+Class-Based Middleware Model
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The middleware model has been changed to a class-based model. This allows for
+more flexibility in implementation, for asynchronous processing for example. Previously,
+the middleware were functions that tightly wrapped the `make_request` function of the
```
Comment (pacrob) [inline docs/v7_migration.rst]: I think these function names should be double-ticks for code font, no?

File: docs/internals.rst [hunk 3]
```diff
@@ -117,66 +117,187 @@ You can set a new list of middlewares by assigning to ``provider.middlewares``,
...
+As with previous versions of the library, wrapping the ``make_request`` method of a
+provider is possible. If you wish to prevent making a call under certain conditions,
+for example, you can override the ``_wrap_make_request`` method. This allows for
+defining pre-request processing, skipping or making the request under certain
+conditions, as well as response processing before passing it to the next middleware.
+The order of operations still passes all pre-request processing down the middlewares
+before making the request, and then passes the response back up the middleware stack
+for processing. The next middleware on the stack is essentially the "make_request"
+method until we reach the end of the middlewares and the request is made by the actual
+`make_request` method of the provider. The response is then passed back up the
```
Comment (pacrob) [inline docs/internals.rst]: double-tick

File: docs/internals.rst
Comment (pacrob) [inline docs/internals.rst:1]: Line 13 says:
These can be *global* operating on all providers or specific to one provider.

I don't think this is still true - we don't have groups of providers anymore, right?

Comment (pacrob) [reply docs/internals.rst:1]: The diagram in *Request Lifecycle* shows both Global and Provider Middlewares

File: docs/internals.rst [hunk 4]
```diff
@@ -117,66 +117,187 @@ You can set a new list of middlewares by assigning to ``provider.middlewares``,
...
 
-It is also possible to implement middlewares as a class.
+        async def async_request_processor(self, method, params):
+            return (method, params)
 
+        async def async_response_processor(self, method, response):
+            return response
+
+
+.. code-block:: python
```
Comment (reedsa) [inline docs/internals.rst]: Extraneous code block
```suggestion
```

**Review by reedsa — APPROVED** [review]
Had one tiny suggestion but looks good.

**Comment by fselmo** [comment]
@pacrob Thanks for the comments. Updated 👀 .

**Review by pacrob — APPROVED** [review]
lgtm!

File: docs/examples.rst [hunk 5]
```diff
@@ -181,7 +181,7 @@ There are a few options for sending transactions:
 - :meth:`~web3.eth.Eth.send_transaction`
 - :meth:`~web3.eth.Eth.send_raw_transaction`
 - Calling :meth:`~web3.contract.ContractFunction.transact` on a contract function
-- Utilizing :meth:`~web3.middleware.construct_sign_and_send_raw_middleware`
+- Configuring the ``build`` method for the :class:`~web3.middleware.SignAndSendRawMiddlewareBuilder`
```
Comment (kclowes) [inline docs/examples.rst]: nit: what do you think about something more human readable here? Maybe just `Configuring sign and send middleware` or something like that? With a link to the docs on how to do that

File: docs/internals.rst [hunk 6]
```diff
@@ -117,66 +111,185 @@ You can set a new list of middlewares by assigning to ``provider.middlewares``,
...
+
+        async def async_request_processor(self, method, params):
+            return (method, params)
+
+        async def async_response_processor(self, method, response):
+            return response
 
-It is also possible to implement middlewares as a class.
 
+As with previous versions of the library, wrapping the ``make_request`` method of a
```
Comment (kclowes) [inline docs/internals.rst]: I would get rid of the 'As with previous versions of the library'. I don't think it's relevant here moving forward.

File: docs/middleware.rst [hunk 7]
```diff
@@ -26,33 +20,20 @@ More information is available in the "Internals: :ref:`internals__middlewares`"
...
-
-* ``gas_price_strategy``
-* ``name_to_address``
-* ``attrdict``
-* ``validation``
-* ``abi``
-* ``gas_estimate``
-
-Async middlewares include:
+The following middleware are added by default if you don't add any:
```
Comment (kclowes) [inline docs/middleware.rst]: middlewares?

**Review by kclowes — APPROVED** [review]
Looks good to me! I left a few nits but nothing substantial. ⛵ I think we're slightly inconsistent with the use of middleware/middlewares for plural middleware, but I also don't think it matters too much. There were a few instances in the file I left the `middlewares?` comment on, but feel free to take it or leave it!

**Comment by fselmo** [comment]
Sounds good, I've resolved all the outstanding questions here and also changed `_wrap_make_request` -> `wrap_make_request` for `Web3Middleware` classes since we instruct how to override that as well. It's an actually useful method to override and utilize so the underscore just adds confusion.

I also rebased to fix conflicts in tests.

**Pull request merged**